OpenRA.Mods.Common/Traits/EntersTunnels.cs

```diff
@@ -0,0 +1,133 @@
+	{
+		public readonly string EnterCursor = "enter";
+		public readonly string EnterBlockedCursor = "enter-blocked";
+
+		[VoiceReference] public readonly string Voice = "Action";
+
+		public object Create(ActorInitializer init) { return new EntersTunnels(init.Self, this); }
+	}
+
+	public class EntersTunnels : IIssueOrder, IResolveOrder, IOrderVoice
+	{
+		readonly EntersTunnelsInfo info;
+		readonly IMove move;
+
+		public EntersTunnels(Actor self, EntersTunnelsInfo info)
+		{
+			this.info = info;
+			move = self.Trait<IMove>();
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get
+			{
+				yield return new EnterTunnelOrderTargeter(info);
+			}
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (order.OrderID != "EnterTunnel")
+				return null;
+
+			if (target.Type == TargetType.FrozenActor)
+				return new Order(order.OrderID, self, queued) { ExtraData = target.FrozenActor.ID, SuppressVisualFeedback = true };
+
+			return new Order(order.OrderID, self, queued) { TargetActor = target.Actor, SuppressVisualFeedback = true };
+		}
+
+		public string VoicePhraseForOrder(Actor self, Order order)
+		{
+			return order.OrderString == "EnterTunnel" ? info.Voice : null;
+		}
+
+		public void ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString != "EnterTunnel")
+				return;
+
+			var target = self.ResolveFrozenActorOrder(order, Color.Red);
+			if (target.Type != TargetType.Actor)
+				return;
+
+			var tunnel = target.Actor.TraitOrDefault<TunnelEntrance>();
+			if (!tunnel.Exit.HasValue)
+				return;
+
+			if (!order.Queued)
+				self.CancelActivity();
+
+			self.SetTargetLine(Target.FromCell(self.World, tunnel.Exit.Value), Color.Green);
+			self.QueueActivity(move.MoveTo(tunnel.Entrance, tunnel.NearEnough));
+			self.QueueActivity(move.MoveTo(tunnel.Exit.Value, tunnel.NearEnough));
+		}
+
+		class EnterTunnelOrderTargeter : UnitOrderTargeter
+		{
+			readonly EntersTunnelsInfo info;
+
+			public EnterTunnelOrderTargeter(EntersTunnelsInfo info)
+				: base("EnterTunnel", 6, info.EnterCursor, true, true)
+			{
+				this.info = info;
+			}
+
+			public override bool CanTargetActor(Actor self, Actor target, TargetModifiers modifiers, ref string cursor)
+			{
+				if (target.IsDead)
+					return false;
+
+				var tunnel = target.TraitOrDefault<TunnelEntrance>();
+				if (tunnel == null)
+					return false;
+
+				// HACK: The engine does not support HiddenUnderFog combined with buildings that use the "_" footprint
+				// We therefore have to use AlwaysVisible and then force-disable interacting with the entrance under shroud
+				var buildingInfo = target.Info.TraitInfoOrDefault<BuildingInfo>();
+				if (buildingInfo != null)
+				{
+					var footprint = FootprintUtils.PathableTiles(target.Info.Name, buildingInfo, target.Location);
+					if (footprint.All(c => self.World.ShroudObscures(c)))
+						return false;
+				}
+
+				if (!tunnel.Exit.HasValue)
```

Comment:
Moving this check above the `BuildingInfo` check could save us those lookups (trait, footprint, shroud) in some cases.